### PR TITLE
IAttention FP8

### LINKED ISSUE #4200: 🐛 [Bug] Torch-TRT does not translate softmax quantizer generated by modelopt fp8 mha quantization
##  Bug Description

The modelopt inserts QDQ pairs for BMM1/softmax/BMM2 to follow the TRT's fp8 MHA pattern however it seems Torch-trt is omitting the softmax quantizer and it caused no fused mha picked up by TRT

## To Reproduce

1. Launch the nvidia pytorch container: `nvcr.io/nvidia/pytorch:26.03-py3`
2. Install transformers: `pip install transformers`
3. Install modelopt nightly: `pip install --upgrade "git+https://github.com/NVIDIA/Model-Optimizer.git@main"`
4. Run the attached scripts

[vit_fp8_mha_qdq_inspect.log](https://github.com/user-attachments/files/26944094/vit_fp8_mha_qdq_inspect.log)
[vit_fp8_mha_qdq_inspect.py](https://github.com/user-attachments/files/26944093/vit_fp8_mha_qdq_inspect.py)

## Expected behavior

The softmax quantizers get converted to TRT qdq along with other bmm quantizers

## Environment

See steps to reproduce

## Additional context

@narendasan filed per discussion

### PR DESCRIPTION
**Pull request by narendasan**
# Description

Prototype support for FP8 normalization scale in IAttention layer

Structurally I think this approach is reasonable, to extract relevant graph level info and bake it into metadata that gets consumed in the converter. I dont think we need the softmax qdq pass necessarily if model opt inserts this for us. 

cc: @nvyihengz, @yizhuoz004 

Fixes #4200, #4167 

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/attention.py [hunk 1]
```diff
@@ -16,6 +18,35 @@
...
+    if ctx.current_node is None:
+        return False
+    scale_val = ctx.current_node.meta.get("_fp8_softmax_scale")
+    if scale_val is None:
+        return False
+    scale_tensor = get_trt_tensor(
+        ctx,
+        torch.tensor(scale_val, dtype=torch.float32),
+        name + "_softmax_fp8_scale",
+        dtype=torch.float32,
```
Comment (yizhuoz004) [inline py/torch_tensorrt/dynamo/conversion/impl/attention.py]: dtype needs to match the pre-quant QKV dtype. otherwise TRT compilatio will fail on some platforms

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/impl/attention.py]: Do you know where we can fetch this info?

Comment (yizhuoz004) [reply py/torch_tensorrt/dynamo/conversion/impl/attention.py]: https://github.com/pytorch/TensorRT/pull/4204/changes/7f0d61ceb114ebf31c156feea039dbec42f2c407  I pulled the attention layer's output tensor's dtype.

File: py/torch_tensorrt/dynamo/conversion/_ConversionContext.py
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/_ConversionContext.py:19]: `current_node` is needed in doc

File: py/torch_tensorrt/dynamo/lowering/passes/insert_fp8_softmax_qdq.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+_FP8_E4M3_SOFTMAX_AMAX = 1.0
+_SOFTMAX_TARGETS = {
+    torch.ops.aten._softmax.default,
+    torch.ops.aten.softmax.int,
+}
+_MATMUL_TARGETS = {
+    torch.ops.aten.matmul.default,
+    torch.ops.aten.bmm.default,
+}
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/lowering/passes/insert_fp8_softmax_qdq.py]: We have targets:
```
@dynamo_tensorrt_converter(torch.ops.aten.matmul, supports_dynamic_shapes=True)
@dynamo_tensorrt_converter(torch.ops.aten.matmul.default, supports_dynamic_shapes=True)
@dynamo_tensorrt_converter(torch.ops.aten.dot.default, supports_dynamic_shapes=True)
@dynamo_tensorrt_converter(torch.ops.aten.mm.default, supports_dynamic_shapes=True)
@dynamo_tensorrt_converter(torch.ops.aten.mv.default, supports_dynamic_shapes=True)
@dynamo_tensorrt_converter(torch.ops.aten.bmm.default, supports_dynamic_shapes=True)
```
whether other targets are needed here?

**Review by zewenli98 — COMMENTED** [review]
I think this approach works for both decompose and not decompose attention paths. Ideally, this work should be done on ModelOpt side I think

**Comment by narendasan** [comment]
@zewenli98 

Can you try this patch to the model code and see if it helps?

```py
from modelopt.torch.quantization.plugins.attention import register_attention_for_kv_quant

register_attention_for_kv_quant(SimpleSDPAModel)
```

```py


```python
import copy
FP8_MHA_CFG = copy.deepcopy(mtq.FP8_DEFAULT_CFG)
FP8_MHA_CFG["quant_cfg"].extend([
    {"quantizer_name": "*q_bmm_quantizer", "cfg": {"num_bits": (4, 3), "axis": None}},
    {"quantizer_name": "*k_bmm_quantizer", "cfg": {"num_bits": (4, 3), "axis": None}},
    {"quantizer_name": "*v_bmm_quantizer", "cfg": {"num_bits": (4, 3), "axis": None}},
])
```

**Comment by zewenli98** [comment]
@narendasan Yes, the patch works for SimpleSDPAModel example

File: py/torch_tensorrt/dynamo/lowering/passes/insert_fp8_softmax_qdq.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
+import logging
+from typing import Optional
+
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/lowering/passes/insert_fp8_softmax_qdq.py]: This entire file can be removed because [the ModelOpt PR ](https://github.com/NVIDIA/Model-Optimizer/pull/1569)handles the decomposed path for us.

File: tests/py/dynamo/models/test_models_export.py [hunk 4]
```diff
@@ -636,3 +636,300 @@ def calibrate_loop(model):
...
+        cos > 0.99
+    ), f"FP8 MHA output deviates from PyTorch reference: cosine_similarity={cos}"
+
+
+@unittest.skipIf(
+    torch.cuda.get_device_capability() < (8, 9),
+    "FP8 quantization requires compute capability 8.9 or later",
+)
+@pytest.mark.unit
+def test_fp8_mha_fused_kernel_decomposed(ir):
```
Comment (zewenli98) [inline tests/py/dynamo/models/test_models_export.py]: This test can be deleted accordingly.

**Review by zewenli98 — COMMENTED** [review]
ModelOpt handles the decomposed path; Torch-TRT only needs to handle IAttention path.

**Comment by zewenli98** [comment]
@narendasan Can you take a final pass? I have verified this PR handles FP8 IAttention path and ModelOpt handles FP8 decomposed path.

**Comment by narendasan** [comment]
@zewenli98 seems fine to me.

**Review by narendasan — COMMENTED** [review]
LGTM
